### PR TITLE
fix: scope the internal expiration sweeps to their own containers

### PR DESCRIPTION
**Pull request by bourgeoa**
This component is using the work made by @jeswr for solidcommunity.net implementation

## Problem

CSS keeps its internal key-value stores under `/.internal/` — login cookies,
forgot-password requests, WebID ownership tokens, OIDC adapter state. Each is
wrapped in an expiring storage whose periodic sweep deletes expired entries, and
the stock wrapper chain delegates `entries()` to a `JsonResourceStorage` rooted
at `/.internal/`: the keys are filtered only **after** the whole internal tree has
been walked recursively.

With a large account base that walk is expensive (every account's data and index,
the IDP records, the lock files), the four stores are all created at startup with
the same interval so their sweeps fire in the same instant, and each sweep deletes
everything expired in one unbounded `Promise.all`, with every delete going through
the locking store.

Observed on a server with ~1000 accounts: 95–100 % CPU with no client traffic,
latency spikes while a sweep ran, and sweep deletes competing with requests for
the resource locks.

## Fix

Two Pivot components plus configuration; CSS itself is untouched.

* `ScopedJsonResourceStorage` — a `JsonResourceStorage` whose `entries()` starts in
  a descendant container. Key mapping, key hashing, identifiers and the on-disk
  layout are unchanged, so it slots into the existing wrapper chain and no stored
  data moves.
* `PivotExpiringStorage` — the same expiry semantics as `WrappedExpiringStorage`,
  plus a jitter on the sweep interval (default 0.15 of the timeout), deletions in
  bounded batches (default 32) and a `finalize()` that clears the sweep timer.
* `config/pivot-scoped-sweeps.json` — overrides the four stores to
  `PivotExpiringStorage → ContainerPathStorage → MaxKeyLengthStorage →
  ScopedJsonResourceStorage(entryContainer = the store's own container)`,
  redeclares `IdpAdapterFactory` (the stock 7.2.0 config declares its storage
  inline, without an `@id` to override) and registers the four stores in
  `urn:solid-server:default:Finalizer` so the timers stop on shutdown.

Imported by `prod.json`, `suffix.json`, `dev-http-suffix.json` and
`dev-http-subdomain.json`; `test.json` is excluded (it has no accounts, so the
stores it would override do not exist there). The sweep interval policy is
unchanged (`timeout: 1` minute — previously inline overrides in `prod.json` /
`suffix.json`, now part of the new file).

The four chains keep the stock locked ResourceStore as their source (the ResourceStore_Locking middleware), so internal reads and writes take the same per-resource locks as before this PR; a sweep only changes where it enumerates and when it runs. /.internal/ stays hidden from clients.

## Verification

* `npx jest` — 6 suites / 67 tests, including new unit coverage for both
  components (scoped enumeration keeps root-relative keys; long keys are still
  hashed and deleted through the wrapper stack; jitter, `unref`, batched deletes,
  `finalize`) and a configuration test that instantiates all four stores from
  `prod.json` and asserts the chains and containers.
* `npm run build` (tsc + Components.js generator) clean.
* Instantiated the whole app graph (`urn:solid-server:default:App`) and
  `urn:solid-server:default:Finalizer` (5 handlers: server initializer + the four
  expiring stores) with Components.js.
* The scoped walk was measured in production on a ~1000-account server: CPU from
  over 95 % with no traffic down to below 1 %, latency flat.

## Documentation

`CLEAR-LOCK.md` — implementation, verification, decisions and limitations.

## Notes

* Pivot keeps its own copy of the expiring storage body (three documented deltas).
  A container-scoped `entries()` in CSS, or an `@id` for the adapter storage, would
  make both components unnecessary.
* The lock configuration (file vs Redis, retry bounds) is orthogonal and unchanged.

### CODE REVIEW

**Comment by jeswr** [comment]
Is it possible to directly use the components proposed upstream, either installing the branch on https://github.com/CommunitySolidServer/CommunitySolidServer or from the alpha release I made? We should avoid having similar-but-different code as much as possible as it is going to create a maintenance nightmare for us.

**Comment by bourgeoa** [comment]
@jeswr I don't want maintenance. the component shall be deleted as soon one is merged in published CSS or any other way if chosen. There is no data change nor in `pods` nor in `.internals`

That's also the reason why 
- I do not want to publish as CSS component outside Pivot
- I also removed/closed my proposal on CSS to only keep yours

Users are in a situation where their servers are not performing well

Your proposals:
- We are not in situation to create branches in or publish from CSS.
- using your fork created an install issue on the test server. I had to use `npm install --force`

File: config/pivot-scoped-sweeps.json [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        "@type": "PivotExpiringStorage",
+        "timeout": 1,
+        "source": {
+          "@type": "ContainerPathStorage",
+          "relativePath": "/accounts/cookies/",
+          "source": {
+            "@type": "MaxKeyLengthStorage",
+            "source": {
+              "@type": "ScopedJsonResourceStorage",
+              "source": { "@id": "urn:solid-server:default:ResourceStore_Backend" },
```
[the remark this answers is not in the document]
Comment (bourgeoa) [reply config/pivot-scoped-sweeps.json]: Fixed: all four chains source `urn:solid-server:default:ResourceStore` again, so internal reads and writes take the same per-resource locks as the stock `KeyValueStorage`; only `BackendKeyValueStorage` uses the backend directly. The scoping (enumeration starts in the store's own container) and the batched deletes are unchanged, so a sweep still reads only its own subtree and its deletes take the same per-entry lock as before this PR. `CLEAR-LOCK.md` and the configuration test were updated accordingly.

File: src/storage/PivotExpiringStorage.ts [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+    this.source = source;
+    this.batchSize = batchSize;
+    const period = timeout * 60 * 1000;
+    const jitterMs = Math.floor(Math.random() * period * jitter);
+    this.timer = setSafeInterval(
+      this.logger,
+      'Failed to remove expired entries',
+      this.removeExpiredEntries.bind(this),
+      period + jitterMs,
+    );
```
[the remark this answers is not in the document]
Comment (bourgeoa) [reply src/storage/PivotExpiringStorage.ts]: Fixed: the sweep is no longer armed on an interval. The next run is scheduled with `setTimeout` only after the current sweep has settled, so a cleanup that takes longer than the timeout cannot start a second enumeration; the jitter is recomputed for every run, failures are logged instead of thrown out of the timer, the timer stays `unref`'d and `finalize()` clears the pending run. Tests cover the no-overlap behaviour and finalizing while a sweep is running.

File: src/storage/ScopedJsonResourceStorage.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+ * on-disk layout exactly the same, but starts the enumeration at `entryContainer` instead of the
+ * storage root, so a sweep only reads its own subtree.
+ */
+export class ScopedJsonResourceStorage<T> extends JsonResourceStorage<T> {
+  private readonly entryContainer: string;
+
+  public constructor(source: ResourceStore, baseUrl: string, container: string, entryContainer: string) {
+    super(source, baseUrl, container);
+    this.entryContainer = ensureTrailingSlash(joinUrl(baseUrl, entryContainer));
+    if (!this.entryContainer.startsWith(this.container)) {
```
[the remark this answers is not in the document]
Comment (bourgeoa) [reply src/storage/ScopedJsonResourceStorage.ts]: Made explicit and covered by a test: the check now normalises the root (`ensureTrailingSlash(this.container)`) before comparing, so a storage rooted at `/.internal/accounts/` rejects `/.internal/accounts-evil/`. `JsonResourceStorage` already stores its root with a trailing slash, so the previous form was segment-aware in practice too - the explicit normalisation keeps it that way regardless of how the root is provided, and the new unit test pins the sibling-prefix case.

File: config/pivot-scoped-sweeps.json [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+        "@type": "PivotExpiringStorage",
+        "timeout": 1,
+        "source": {
+          "@type": "ContainerPathStorage",
+          "relativePath": "/accounts/cookies/",
+          "source": {
+            "@type": "MaxKeyLengthStorage",
+            "source": {
+              "@type": "ScopedJsonResourceStorage",
+              "source": { "@id": "urn:solid-server:default:ResourceStore" },
```
[the remark this answers is not in the document]
Comment (bourgeoa) [reply config/pivot-scoped-sweeps.json]: Fixed: `has()` compares the retrieved value with `undefined` instead of coercing it, so falsy payloads (`''`, `0`, `false`) are present for both `get()` and `has()`. The default `WrappedExpiringStorage` has the same coercion; `PivotExpiringStorage` is Pivot's own copy, so it now follows the contract. Covered by a unit test with an empty-string payload.

File: src/storage/PivotExpiringStorage.ts [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+    timeout = 60,
+    jitter = 0.15,
+    batchSize = 32,
+  ) {
+    if (!Number.isSafeInteger(batchSize) || batchSize < 1) {
+      throw new TypeError('The expired-entry deletion batch size must be a positive integer.');
+    }
+    this.source = source;
+    this.timeout = timeout;
+    this.jitter = jitter;
```
[the remark this answers is not in the document]
Comment (bourgeoa) [reply src/storage/PivotExpiringStorage.ts]: Added. The storage now validates its timing at construction: timeout must be a positive finite number and jitter a non-negative finite number, otherwise a TypeError is thrown. That makes a misconfigured value fail loudly instead of being clamped to ~1 ms by setTimeout and turning the sweep into a busy loop. Unit tests cover 0, negative, NaN and Infinity for both parameters.

File: src/storage/PivotExpiringStorage.ts [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+    const expired: TKey[] = [];
+    for await (const [ key, value ] of this.source.entries()) {
+      const { expires } = this.toData(value);
+      if (this.isExpired(expires)) {
+        expired.push(key);
+      }
+    }
+    for (let index = 0; index < expired.length; index += this.batchSize) {
+      await Promise.all(expired.slice(index, index + this.batchSize)
+        .map(async(key): Promise<boolean> => this.source.delete(key)));
```
[the remark this answers is not in the document]
Comment (bourgeoa) [reply src/storage/PivotExpiringStorage.ts]: Fixed. A batch now collects its delete promises with Promise.allSettled, so every delete of the batch settles before a rejection is rethrown and before the sweep reschedules itself. A unit test rejects one delete while a sibling is still pending and asserts that neither the next batch nor the next scheduled run happens until that sibling finishes.

File: CLEAR-LOCK.md [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+
+A `JsonResourceStorage` whose enumeration starts in a descendant container. Everything
+else — key mapping, key hashing, identifiers, file layout — is inherited unchanged:
+
+```ts
+export class ScopedJsonResourceStorage<T> extends JsonResourceStorage<T> {
+  public constructor(source: ResourceStore, baseUrl: string, container: string, entryContainer: string) {
+    super(source, baseUrl, container);
+    this.entryContainer = ensureTrailingSlash(joinUrl(baseUrl, entryContainer));
+    if (!this.entryContainer.startsWith(this.container)) {
```
[the remark this answers is not in the document]
Comment (bourgeoa) [reply CLEAR-LOCK.md]: Updated. The constructor snippet in CLEAR-LOCK.md now uses the segment-aware entryContainer.startsWith(ensureTrailingSlash(this.container)) check, and the batch snippet reflects the settle-then-fail implementation.

File: src/storage/PivotExpiringStorage.ts [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+    batchSize = 32,
+  ) {
+    if (!Number.isSafeInteger(batchSize) || batchSize < 1) {
+      throw new TypeError('The expired-entry deletion batch size must be a positive integer.');
+    }
+    if (!Number.isFinite(timeout) || timeout <= 0) {
+      throw new TypeError('The sweep timeout must be a positive finite number of minutes.');
+    }
+    if (!Number.isFinite(jitter) || jitter < 0) {
+      throw new TypeError('The sweep jitter must be a non-negative finite number.');
```
[the remark this answers is not in the document]
Comment (bourgeoa) [reply src/storage/PivotExpiringStorage.ts]: Good catch. The validation now bounds the delay, not just its finiteness: the constructor computes the worst-case jittered delay timeout * 60_000 * (1 + jitter) and rejects it with a TypeError when it exceeds the maximum setTimeout delay (2 ** 31 - 1 ms, about 24.8 days), since Node.js clamps anything above that to 1 ms and the sweep would become the busy loop this validation exists to prevent. Tests cover a huge finite timeout (1e9), a huge finite jitter, one minute over the limit (35792) and a mid-range timeout whose jitter pushes it over (30000, 0.5), plus a boundary case confirming the largest supported delay (35791 minutes) is still accepted. I chose rejection over chunked long-delay scheduling on purpose: a delay of that magnitude is a configuration mistake, and chunking would add scheduling drift and extra state for every normal deployment.
